edb/schema/pointers.py

```diff
@@ -241,30 +255,50 @@ def get_verbosename(self, schema, *, with_parent: bool=False) -> str:
     def is_scalar(self) -> bool:
         return False
 
-    def material_type(self, schema):
+    def material_type(self, schema: s_schema.Schema) -> Pointer:
         non_derived_parent = self.get_nearest_non_derived_parent(schema)
         if non_derived_parent.generic(schema):
             return self
         else:
             return non_derived_parent
 
-    def get_near_endpoint(self, schema, direction):
+    def get_near_endpoint(
+        self,
+        schema: s_schema.Schema,
+        direction: PointerDirection,
+    ) -> Optional[so.Object]:
```

Comment:
I have doubts that I am annotating properly `get_near_endpoint` and `get_far_endpoint` (Optional[so.Object] instead of Optional[s_types.Type]), but since in the current code, `source` and `target` can be swapped depending on the pointer direction; and since they are already annotated as schema fields (`Object` and `Type), I am not sure whether SchemaFields should be modified, or these functions return type annotation made more high-level as I opted for.

Reply:
If the pointer is a link property, then the near endpoint would be a `Link` instance, which isn't a `Type`, so this is correct.